Minio/Credentials/IAMAWSProvider.cs

```diff
@@ -0,0 +1,265 @@
+            return this.Credentials;
+        }
+        public async Task<string> GetIamRoleNameAsync(Uri url)
+        {
+            this.Validate();
+            string[] roleNames = null;
+            RestRequest request = new RestRequest(url.ToString(), Method.GET);
+            var response = await this.Minio_Client.ExecuteAsync(Enumerable.Empty<ApiResponseErrorHandlingDelegate>(), request);
+            if (string.IsNullOrWhiteSpace(response.Content) ||
+                    !HttpStatusCode.OK.Equals(response.StatusCode))
+            {
+                throw new CredentialsProviderException("IAMAWSProvider", "Credential Get operation failed with HTTP Status code: " + response.StatusCode);
+            }
+            roleNames = response.Content.Split('\n');
+            if (roleNames.Length <= 0)
+            {
+                throw new CredentialsProviderException("IAMAWSProvider", "No IAM roles are attached to AWS service at "+ url.ToString());
+            }
+            int index = 0;
+            foreach (var item in roleNames)
+            {
+                roleNames[index++] = item.Trim();
+            }
+            return roleNames[0];
+        }
+
+        public async Task<Uri> GetIamRoleNamedURL()
+        {
+            this.Validate();
+            Uri url = this.CustomEndPoint;
+            string newUrlStr = null;
+            if (url == null || string.IsNullOrWhiteSpace(url.Authority))
+            {
+                url = new Uri("http://169.254.169.254/latest/meta-data/iam/security-credentials/");
+                newUrlStr = "http://169.254.169.254/latest/meta-data/iam/security-credentials/";
+            }
+            else
+            {
+                var urlStr = url.Scheme + "://" + url.Authority + "/latest/meta-data/iam/security-credentials/";
+                url = new Uri(urlStr);
+                newUrlStr = urlStr;
+            }
+            string roleName = await this.GetIamRoleNameAsync(url);
+            newUrlStr += roleName;
+            return new Uri(newUrlStr);
+        }
+
+        public IAMAWSProvider()
+        {
+            this.Minio_Client = null;
+        }
+
+        public IAMAWSProvider WithMinioClient(MinioClient minio)
+        {
+            this.Minio_Client = minio;
+            if (this.Credentials == null ||
+                string.IsNullOrWhiteSpace(this.Credentials.AccessKey) || string.IsNullOrWhiteSpace(this.Credentials.SecretKey))
+            {
+                this.Credentials = this.GetCredentialsAsync().GetAwaiter().GetResult();
+            }
+            return this;
+        }
+
+        public IAMAWSProvider(string endpoint, HttpClient client)
+        {
+            if (string.IsNullOrWhiteSpace(endpoint))
+            {
+                throw new ArgumentNullException("Endpoint field " + nameof(CustomEndPoint) + " cannot be null or empty.");
+            }
+            this.CustomEndPoint = new Uri(endpoint);
```

Comment:
the HttpClient is not being used here?

Reply:
Thanks for pointing out. I will remove the constructor.